devlib/instrument/gem5stats.py

```diff
@@ -0,0 +1,152 @@
+
+    This instrument parses the stats file produced by Gem5 and extracts power
+    values. More information on the "stats.txt" file structure can be found 
+    here:
+    
+    http://pages.cs.wisc.edu/~david/courses/cs752/Spring2015/gem5-tutorial/part1/gem5_stats.html
+    
+    This instrument also implements Gem5's ROIs (Regions of Interest). A ROI is
+    an annotation on a time interval marked by start and end cursors. This enables
+    power analysis on a specific time window which spans over the execution of
+    a workload for example. Up to 8 ROIs can be used, with or without 
+    overlapping.
+    '''
+
+    mode = CONTINUOUS
+    dump_head='---------- Begin Simulation Statistics ----------'
+    dump_tail='---------- End Simulation Statistics   ----------'
+    
+    def __init__(self, target, channel_map, stats_file_path):
+        super(Gem5StatsInstrument, self).__init__(target)
+        
+        # Duration since last dump is indexed by 'sim_seconds' in stats.txt
+        self.add_channel('Duration', 'time', 'sim_seconds')
+        # Track power only for now
+        for f in channel_map:
+            self.add_channel(f, 'power', channel_map[f])
+        
+        self._stats_file_path = stats_file_path
+        self._roi_state = [False] * 8
+        self.dumps = []
+
+    def start(self, roi=0):
+        if roi > 7 or roi < 0:
+            raise IndexError('Incorrect ROI number')
+        if self._roi_state[roi]:
+            raise TargetError('ROI {} already running'.format(roi))
+        self._roi_state[roi] = True
+        self.target.execute('m5 roistart {}'.format(roi))
+        self.target.execute('m5 dumpresetstats')
+
+    def stop(self, roi=0):
+        if roi > 7 or roi < 0:
+            raise IndexError('Incorrect ROI number')
+        if not self._roi_state[roi]:
+            raise TargetError('ROI {} not running'.format(roi))
+        self._roi_state[roi] = False
+        self.target.execute('m5 dumpresetstats')
+        self.target.execute('m5 roiend {}'.format(roi))
+
+    def get_data(self, outfile, roi=0):
+        if roi > 7 or roi < 0:
+            raise RuntimeError('Incorrect ROI number')
+        self.target.execute('m5 dumpresetstats')
+        with open(self._stats_file_path, 'r') as f:
+            self._parse_stats(f)
+        with open(outfile, 'wb') as wfh:
+            writer = csv.writer(wfh)
+            cur_chans = sorted([c.label for c in self.active_channels])
+            writer.writerow(cur_chans) # headers
+            for s in (self.dumps):
+                if int(s['ROI::{}'.format(roi)])==0: # ignore out-of-roi dumps
+                    continue
+                row = [float(_get_val(s, self.channels[c].name)) for c in cur_chans]
+                writer.writerow(row)
+        return MeasurementsCsv(outfile, self.active_channels)
+    
+    def reset(self, sites=None, kinds=None, channels=None):
+        '''
+        Beware that reset() does **NOT** erase "stats.txt". Doing so while 
```

Comment:
This breaks the API. The API dictates that reset must reset the instrument for a fresh collection session. The calling code may call reset-start-stop-get_data to do something on the target and then immediately perform another reset-start-stop-get_data session, and the stats collected for that must not be influenced by the previous session.

Perhaps this should not be manipulating the raw stats file directly. Access to the stats can be mediated via something like a `get_stats` method of a gem5-specific module, which would track where in the stats file it was last time it was invoked and offset accordingly on each call.